File: Trabalhos Praticos/tp03/TP03Q02 - Lista com Alocação Sequencial em C/Personagem.c
//----------------------------------------------------------------------------------------//
// Avisos
//----------------------------------------------------------------------------------------//
// Olá, esse código provavelmente não está organizado como eu gostaria,
// por conta da urgência das entregas.
// E é normal ter muitas linhas de código comentadas,
// já que é para uma atividade, eu aproveito isso para documentar todo o estudo também.
//----------------------------------------------------------------------------------------//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Lista
{
    char codenomes[10][256]; // Supondo que cada nome não exceda 255 caracteres
} Lista;

void inicializarLista(Lista *lista)
{
    for (int i = 0; i < 10; i++)
    {
        strcpy(lista->codenomes[i], "VAZIO"); // Inicializa todos os nomes como strings vazias
    }
}

typedef struct Personagem
{
    char id[256];
    char name[256];
    Lista alternate_name;
    char house[256];
    char ancestry[256];
    char species[256];
    char patronus[256];
    int hogwartsStaff;   // bool
    int hogwartsStudent; // bool
    char actorName[256];
    int alive;            // bool
    char dateOfBirth[11]; // Formato "dd-mm-yyyy"
    int yearOfBirth;
    char eyeColour[256];
    char gender[256];
    char hairColour[256];
    int wizard; // bool
} Personagem;

typedef struct ListaPersonagem
{
    int tam, maxTam;
    Personagem **array;
} ListaPersonagem;

void inicializarListaPersonagem(ListaPersonagem *lista, int maxTam)
{
    lista->tam = 0;
    lista->maxTam = maxTam;
    lista->array = (Personagem **)malloc(maxTam * sizeof(Personagem *));
}

void imprimirLista(const Lista *lista)
{
    printf("{");
    int i;
    for (i = 0; i < 10; i++)
    {
        if (strcmp(lista->codenomes[i], "VAZIO") != 0)
        {
            printf("%s", lista->codenomes[i]);
            if (i < 9 && strcmp(lista->codenomes[i + 1], "VAZIO") != 0)
            {
                printf(","); // Adiciona uma vírgula apenas se houver próximo elemento
            }
        }
    }
    printf("} ## ");
}

void imprimir(const Personagem *p)
{
    printf("%s ## %s ## ", p->id, p->name);
    imprimirLista(&p->alternate_name);
    printf("%s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %s ## %s ## %s ## %s]\n",
           p->house, p->ancestry, p->species, p->patronus,
           p->hogwartsStaff ? "true" : "false",
           p->hogwartsStudent ? "true" : "false",
           p->actorName,
           p->alive ? "true" : "false",
           p->dateOfBirth,
           p->yearOfBirth,
           p->eyeColour, p->gender, p->hairColour,
           p->wizard ? "true" : "false");
}

int toUSbool(const char *frase)
{
    return strstr(frase, "VERDADEIRO") != NULL;
}

void limparCaracteresEspeciais(char *string)
{
    int i, j;
    for (i = 0, j = 0; string[i] != '\0'; i++)
    {
        if (string[i] != '\'' && string[i] != '[' && string[i] != ']')
        {
            string[j++] = string[i];
        }
    }
    string[j] = '\0';
}

void processarNomesAlternativos(char *nomesAlternativos, Lista *lista)
{
    inicializarLista(lista);
    char *nome = strtok(nomesAlternativos, ",");
    int i = 0;
    while (nome != NULL && i < 10)
    {
        limparCaracteresEspeciais(nome);
        strcpy(lista->codenomes[i++], nome);
        nome = strtok(NULL, ",");
    }
}

void ler(Personagem *p, char *ID)
{
    // FILE *file = fopen("./characters.csv", "r");
    FILE *file = fopen("/tmp/characters.csv", "r");
    if (!file)
    {
        perror("Erro ao abrir o arquivo");
        return;
    }

    char linha[1024];

    while (fgets(linha, sizeof(linha), file))
    {
        char vetor[18][256];
        int index = 0;
        char *token = linha;
        char *next_token;

        while (token)
        {
            next_token = strchr(token, ';');
            if (next_token)
            {
                *next_token = '\0';
                strcpy(vetor[index++], token);
                token = next_token + 1;
            }
            else
            {
                strcpy(vetor[index++], token);
                break;
            }
        }

        if (strcmp(ID, vetor[0]) == 0)
        {
            strcpy(p->id, vetor[0]);
            strcpy(p->name, vetor[1]);
            processarNomesAlternativos(vetor[2], &p->alternate_name);
            strcpy(p->house, vetor[3]);
            strcpy(p->ancestry, vetor[4]);
            strcpy(p->species, vetor[5]);
            strcpy(p->patronus, vetor[6]);
            p->hogwartsStaff = toUSbool(vetor[7]);
            p->hogwartsStudent = toUSbool(vetor[8]);
            strcpy(p->actorName, vetor[9]);
            p->alive = toUSbool(vetor[10]);
            strcpy(p->dateOfBirth, vetor[12]);
            p->yearOfBirth = atoi(vetor[13]);
            strcpy(p->eyeColour, vetor[14]);
            strcpy(p->gender, vetor[15]);
            strcpy(p->hairColour, vetor[16]);
            p->wizard = toUSbool(vetor[17]);
            break;
        }
    }
    fclose(file);
}

void inserirInicio(ListaPersonagem *lista, Personagem *p)
{
    if (lista->tam == lista->maxTam)
    {
        printf("Erro: lista cheia.\n");
        return;
    }
    for (int i = lista->tam; i > 0; i--)
    {
        lista->array[i] = lista->array[i - 1];
    }
    lista->array[0] = p;
    lista->tam++;
}

void inserirFim(ListaPersonagem *lista, Personagem *p)
{
    if (lista->tam == lista->maxTam)
    {
        printf("Erro: lista cheia.\n");
        return;
    }
    lista->array[lista->tam++] = p;
}

void inserir(ListaPersonagem *lista, int pos, Personagem *p)
{
    if (lista->tam == lista->maxTam)
    {
        printf("Erro: lista cheia.\n");
        return;
    }
    for (int i = lista->tam; i > pos; i--)
    {
        lista->array[i] = lista->array[i - 1];
    }
    lista->array[pos] = p;
    lista->tam++;
}

Personagem *removerInicio(ListaPersonagem *lista)
{
    if (lista->tam == 0)
    {
        printf("Erro: lista vazia.\n");
        return NULL;
    }
    Personagem *tmp = lista->array[0];
    for (int i = 0; i < lista->tam - 1; i++)
    {
        lista->array[i] = lista->array[i + 1];
    }
    lista->tam--;
    return tmp;
}

Personagem *removerFim(ListaPersonagem *lista)
{
    if (lista->tam == 0)
    {
        printf("Erro: lista vazia.\n");
        return NULL;
    }
    return lista->array[--lista->tam];
}

Personagem *remover(ListaPersonagem *lista, int pos)
{
    if (lista->tam == 0)
    {
        printf("Erro: lista vazia.\n");
        return NULL;
    }
    Personagem *tmp = lista->array[pos];
    for (int i = pos; i < lista->tam - 1; i++)
    {
        lista->array[i] = lista->array[i + 1];
    }
    lista->tam--;
    return tmp;
}

void hubFuncao(char *linha, ListaPersonagem *lista)
{
    char *palavra[3];
    int i = 0;
    char *token = strtok(linha, " ");

    while (token != NULL && i < 3)
    {
        palavra[i++] = token;
        token = strtok(NULL, " ");
    }

    if (strcmp(palavra[0], "II") == 0)
    {
        Personagem *p = (Personagem *)malloc(sizeof(Personagem));
        ler(p, palavra[1]);
        inserirInicio(lista, p);
    }
    else if (strcmp(palavra[0], "IF") == 0)
    {
        Personagem *p = (Personagem *)malloc(sizeof(Personagem));
        ler(p, palavra[1]);
        inserirFim(lista, p);
    }
    else if (strcmp(palavra[0], "I*") == 0)
    {
        Personagem *p = (Personagem *)malloc(sizeof(Personagem));
        ler(p, palavra[2]);
        inserir(lista, atoi(palavra[1]), p);
    }
    else if (strcmp(palavra[0], "RI") == 0)
    {
        Personagem *p = removerInicio(lista);
        if (p)
            printf("(R) %s\n", p->name);
        free(p);
    }
    else if (strcmp(palavra[0], "RF") == 0)
    {
        Personagem *p = removerFim(lista);
        if (p)
            printf("(R) %s\n", p->name);
        free(p);
    }
    else if (strcmp(palavra[0], "R*") == 0)
    {
        Personagem *p = remover(lista, atoi(palavra[1]));
        if (p)
            printf("(R) %s\n", p->name);
        free(p);
    }
}

int main()
{
    ListaPersonagem lista;
    inicializarListaPersonagem(&lista, 300);

    char ID[256];
    // scanf(" %[^\n]", ID);
    scanf(" %[^\r\n]s", ID);

    while (strcmp(ID, "FIM") != 0)
    {
        Personagem *p = (Personagem *)malloc(sizeof(Personagem));
        ler(p, ID);
        inserirFim(&lista, p);
        scanf(" %[^\r\n]s", ID);
    }

    int numEntradas;
    scanf("%d", &numEntradas);
    getchar();
    // while (getchar() != '\n'); // Limpar o buffer do teclado

    // scanf(" %[^\n]d", numEntradas);

    for (int j = 0; j < numEntradas; j++)
    {
        char linha[256];
        // fgets(linha, sizeof(linha), stdin); // Simplesmente me tomou 2h por causa de um /r
        scanf(" %[^\r\n]s", linha);      // scanf mvp
        linha[strcspn(linha, "\n")] = 0; // Remove newline character
        hubFuncao(linha, &lista);
    }

    for (int i = 0; i < lista.tam; i++)
    {
        printf("[%d ## ", i);
        imprimir(lista.array[i]);
        free(lista.array[i]);
    }

    free(lista.array);
    return 0;
}

// cls && gcc Personagem.c && .\a.exe < pub.in > saida.txt

File: Trabalhos Praticos/tp03/TP03Q05 - Lista com Alocação Flexível em Java/Personagem.java
//----------------------------------------------------------------------------------------//
// Avisos
//----------------------------------------------------------------------------------------//
// Olá, esse código provavelmente não está organizado como eu gostaria,
// por conta da urgência das entregas.
// E é normal ter muitas linhas de código comentadas, 
// já que é para uma atividade, eu aproveito isso para documentar todo o estudo também.
//----------------------------------------------------------------------------------------//

import java.util.Scanner;
import java.io.File;
import java.util.Date;
import java.text.SimpleDateFormat;

class CelulaLista {
    Personagem personagem;
    CelulaLista prox;

    CelulaLista() {
        this.personagem = new Personagem();
        this.prox = null;
    }

    CelulaLista(Personagem p) {
        this.personagem = p;
        this.prox = null;
    }
}

class ListaPersonagem {
    private int tam;
    private CelulaLista primeiro, ultimo;

    public ListaPersonagem() {
        primeiro = ultimo = new CelulaLista();
        tam = 0;
    }

    public int getTam() {
        return this.tam;
    }

    public void imprimir() {
        if (this.tam == 0) {
            System.out.println("Erro: lista vazia.");
            return;
        }

        int cont = 0;
        for (CelulaLista i = this.primeiro.prox; i != null; i = i.prox) {
            System.out.print("[" + cont++ + " ## ");
            i.personagem.imprimir();
        }
    }

    // ---------- Funcoes INICIO -----------//

    public void InserirInicio(Personagem personagem) {
        CelulaLista tmp = new CelulaLista(personagem);
        tmp.prox = primeiro.prox;
        primeiro.prox = tmp;

        if (primeiro == ultimo)
            ultimo = tmp;
        tmp = null;
        this.tam++;
    }

    public void InserirFim(Personagem personagem) {
        CelulaLista tmp = new CelulaLista(personagem);
        ultimo.prox = tmp;
        ultimo = ultimo.prox;
        this.tam++;
    }

    public void Inserir(String posSTRING, Personagem personagem) throws Exception {
        int pos = Integer.parseInt(posSTRING);
        if (pos < 0 || pos > this.tam) {
            throw new Exception("Erro: posicao inexistente");
        } else if (pos == 0) {
            InserirInicio(personagem);
        } else if (pos == this.tam) {
            InserirFim(personagem);
        } else {
            CelulaLista i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;
            CelulaLista tmp = new CelulaLista(personagem);
            tmp.prox = i.prox;
            i.prox = tmp;
            tmp = i = null;
            this.tam++;
        }
    }

    public Personagem RemoverInicio() throws Exception {
        if (this.primeiro == this.ultimo) {
            throw new Exception("Erro: lista vazia.");
        }
        Personagem personagem = primeiro.prox.personagem;
        CelulaLista tmp = primeiro.prox;
        primeiro.prox = primeiro.prox.prox;
        tmp.prox = null;
        tmp = null;
        this.tam--;
        return personagem;
    }

    public Personagem RemoverFim() throws Exception {
        if (this.primeiro == this.ultimo) {
            throw new Exception("Erro: lista vazia.");
        }
        CelulaLista i;
        for (i = primeiro; i.prox != ultimo; i = i.prox)
            ;
        Personagem personagem = i.prox.personagem; // ou (ultimo.elemento)
        ultimo = i;
        ultimo.prox = i = null;
        this.tam--;
        return personagem;
    }

    public Personagem Remover(String posSTRING) throws Exception {
        int pos = Integer.parseInt(posSTRING);
        Personagem personagem = new Personagem();
        if (pos < 0 || pos >= this.tam) {
            throw new Exception("Erro: posicao inexistente");
        } else if (pos == 0) {
            RemoverInicio();
        } else if (pos == this.tam - 1) {
            RemoverFim();
        } else {
            CelulaLista i = primeiro;
            for (int j = 0; j < pos; j++, i = i.prox)
                ;
            CelulaLista tmp = i.prox;
            personagem = tmp.personagem;
            i.prox = tmp.prox;
            tmp = tmp.prox = null;
            this.tam--;
        }
        return personagem;
    }

    // ---------- Funcoes FIM -----------//
}

class Lista {
    String[] codenomes = new String[10];

    public Lista() {
        // Inicializando o array com strings vazias
        for (int i = 0; i < codenomes.length; i++) {
            codenomes[i] = " ";
        }
    }

    public Lista(String nomes) {
        int tam = nomes.length();

        StringBuilder nomesedit = new StringBuilder();

        int j = 0;
        for (int i = 0; i < tam; i++) {
            if (nomes.charAt(i) != '\'' && nomes.charAt(i) != '[' && nomes.charAt(i) != ']') {
                if (nomes.charAt(i) != ',') {
                    nomesedit.append(nomes.charAt(i));
                } else {
                    nomesedit.append(nomes.charAt(i));
                    this.codenomes[j] = nomesedit.toString();
                    nomesedit.setLength(0);
                    j++;
                }
            }
        }
        // }
        this.codenomes[j] = nomesedit.toString();
    }
}

public class Personagem {
    private String id;
    private String name;
    private Lista alternate_name;
    private String house;
    private String ancestry;
    private String species;
    private String patronus;
    private boolean hogwartsStaff;
    private boolean hogwartsStudent;
    private String actorName;
    private boolean alive;
    // alternate_actors
    private Date dateOfBirth;
    private int yearOfBirth;
    private String eyeColour;
    private String gender;
    private String hairColour;
    private boolean wizard;

    public Personagem(String id, String name, Lista alternate_name, String house, String ancestry, String species,
            String patronus, boolean hogwartsStaff, boolean hogwartsStudent, String actorName, boolean alive,
            Date dateOfBirth, int yearOfBirth, String eyeColour, String gender, String hairColour, boolean wizard) {
        this.id = id;
        this.name = name;
        this.alternate_name = alternate_name;
        this.house = house;
        this.ancestry = ancestry;
        this.species = species;
        this.patronus = patronus;
        this.hogwartsStaff = hogwartsStaff;
        this.hogwartsStudent = hogwartsStudent;
        this.actorName = actorName;
        this.alive = alive;
        this.dateOfBirth = dateOfBirth;
        this.yearOfBirth = yearOfBirth;
        this.eyeColour = eyeColour;
        this.gender = gender;
        this.hairColour = hairColour;
        this.wizard = wizard;
    }

    public Personagem() {
        id = " ";
        name = " ";
        alternate_name = new Lista();
        house = " ";
        ancestry = " ";
        species = " ";
        patronus = " ";
        hogwartsStaff = false;
        hogwartsStudent = false;
        actorName = " ";
        alive = false;
        dateOfBirth = null;
        yearOfBirth = -1;
        eyeColour = " ";
        gender = " ";
        hairColour = " ";
        wizard = false;
    }

    // ---------- Metodos GET/SET INICIO -----------//
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Lista getAlternate_name() {
        return alternate_name;
    }

    public void setAlternate_name(Lista alternate_name) {
        this.alternate_name = alternate_name;
    }

    public String getHouse() {
        return house;
    }

    public void setHouse(String house) {
        this.house = house;
    }

    public String getAncestry() {
        return ancestry;
    }

    public void setAncestry(String ancestry) {
        this.ancestry = ancestry;
    }

    public String getSpecies() {
        return species;
    }

    public void setSpecies(String species) {
        this.species = species;
    }

    public String getPatronus() {
        return patronus;
    }

    public void setPatronus(String patronus) {
        this.patronus = patronus;
    }

    public boolean isHogwartsStaff() {
        return hogwartsStaff;
    }

    public void setHogwartsStaff(boolean hogwartsStaff) {
        this.hogwartsStaff = hogwartsStaff;
    }

    public boolean isHogwartsStudent() {
        return hogwartsStudent;
    }

    public void setHogwartsStudent(boolean hogwartsStudent) {
        this.hogwartsStudent = hogwartsStudent;
    }

    public String getActorName() {
        return actorName;
    }

    public void setActorName(String actorName) {
        this.actorName = actorName;
    }

    public boolean isAlive() {
        return alive;
    }

    public void setAlive(boolean alive) {
        this.alive = alive;
    }

    public Date getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(Date dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public int getYearOfBirth() {
        return yearOfBirth;
    }

    public void setYearOfBirth(int yearOfBirth) {
        this.yearOfBirth = yearOfBirth;
    }

    public String getEyeColour() {
        return eyeColour;
    }

    public void setEyeColour(String eyeColour) {
        this.eyeColour = eyeColour;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getHairColour() {
        return hairColour;
    }

    public void setHairColour(String hairColour) {
        this.hairColour = hairColour;
    }

    public boolean isWizard() {
        return wizard;
    }

    public void setWizard(boolean wizard) {
        this.wizard = wizard;
    }

    // ---------- Metodos GET/SET FINAL -----------//
    // ---------- Outros Metodos INICIO -----------//

    public static boolean toUSbool(String frase) {
        if (frase.equals("VERDADEIRO")) {
            return true;
        } else {
            return false;
        }
    }

    public static void imprimirlista(Lista lista) {
        // PROBLEMA NO TAMANHO AQUI
        System.out.print("{");
        for (int i = 0; i < 10; i++) {
            if (lista.codenomes[i] != null) { // Gambiarra?
                System.out.print(lista.codenomes[i]);
            }
        }
        System.out.print("}");
        System.out.print(" ## ");

    }

    public void imprimir() {
        SimpleDateFormat sdfOutput = new SimpleDateFormat("dd-MM-yyyy");

        System.out.print(id + " ## ");
        System.out.print(name + " ## ");
        imprimirlista(alternate_name);
        System.out.print(house + " ## ");
        System.out.print(ancestry + " ## ");
        System.out.print(species + " ## ");
        System.out.print(patronus + " ## ");
        System.out.print(hogwartsStaff + " ## ");
        System.out.print(hogwartsStudent + " ## ");
        System.out.print(actorName + " ## ");
        System.out.print(alive + " ## ");
        System.out.print(sdfOutput.format(dateOfBirth) + " ## ");
        System.out.print(yearOfBirth + " ## ");
        System.out.print(eyeColour + " ## ");
        System.out.print(gender + " ## ");
        System.out.print(hairColour + " ## ");
        System.out.println(wizard + "]");
    }

    public Personagem clone(Personagem personagem) {

        Personagem clone = new Personagem(
                personagem.getId(),
                personagem.getName(),
                personagem.getAlternate_name(),
                personagem.getHouse(),
                personagem.getAncestry(),
                personagem.getSpecies(),
                personagem.getPatronus(),
                personagem.isHogwartsStaff(),
                personagem.isHogwartsStudent(),
                personagem.getActorName(),
                personagem.isAlive(),
                personagem.getDateOfBirth(),
                personagem.getYearOfBirth(),
                personagem.getEyeColour(),
                personagem.getGender(),
                personagem.getHairColour(),
                personagem.isWizard());

        return clone;
    }

    public static Personagem ler(String linhaID) {
        // File file = new File("characters.csv");
        File file = new File("/tmp/characters.csv");

        try {
            Scanner scanner = new Scanner(file);
            scanner.nextLine(); // Descarta a primeira linha
            while (scanner.hasNextLine()) {
                SimpleDateFormat sdfInput = new SimpleDateFormat("dd-MM-yyyy");
                // SimpleDateFormat sdfOutput = new SimpleDateFormat("dd-MM-yyyy");

                String linha = scanner.nextLine();
                String[] vetor = separarlinha(linha);

                String id = vetor[0];
                if (linhaID.equals(id)) {

                    String name = vetor[1];
                    Lista alternate_name = new Lista(vetor[2]);
                    // System.out.println(vetor[2]); // testa
                    // System.out.println(alternate_name.codenomes[3]); // testa
                    String house = vetor[3];
                    String ancestry = vetor[4];
                    String species = vetor[5];
                    String patronus = vetor[6];
                    boolean hogwartsStaff = toUSbool(vetor[7]);
                    boolean hogwartsStudent = toUSbool(vetor[8]);
                    String actorName = vetor[9];
                    boolean alive = toUSbool(vetor[10]);
                    // alternate_actors VETOR 11 NULO
                    Date dateOfBirth = sdfInput.parse(vetor[12]);
                    // String teste = sdfOutput.format(dateOfBirth);
                    int yearOfBirth = Integer.parseInt(vetor[13]);

                    String eyeColour = vetor[14];
                    String gender = vetor[15];
                    String hairColour = vetor[16];

                    boolean wizard = toUSbool(vetor[17]);

                    return new Personagem(id, name, alternate_name, house, ancestry,
                            species,
                            patronus, hogwartsStaff, hogwartsStudent, actorName, alive,
                            dateOfBirth, yearOfBirth, eyeColour, gender, hairColour, wizard);
                }
            }
            // scanner.close(); // Bugado?
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String[] separarlinha(String linha) {
        StringBuilder res = new StringBuilder();
        String[] resfinal = new String[18];
        int tam = linha.length();

        int j = 0;
        for (int i = 0; i < tam; i++) {
            if (linha.charAt(i) != ';') {
                res.append(linha.charAt(i));
            } else {
                resfinal[j] = res.toString();
                res.setLength(0);
                j++;
            }
        }
        resfinal[j] = res.toString();
        return resfinal;
    }

    public static void hubFuncao(String linha, ListaPersonagem lista) throws Exception {

        String palavra[] = linha.split(" ");

        if (palavra[0].equals("II")) {
            lista.InserirInicio(ler(palavra[1]));
        } else if (palavra[0].equals("IF")) {
            lista.InserirFim(ler(palavra[1]));
        } else if (palavra[0].equals("I*")) {
            lista.Inserir(palavra[1], ler(palavra[2]));
        } else if (palavra[0].equals("RI")) {
            System.out.println("(R) " + lista.RemoverInicio().getName());
        } else if (palavra[0].equals("RF")) {
            System.out.println("(R) " + lista.RemoverFim().getName());
        } else if (palavra[0].equals("R*")) {
            System.out.println("(R) " + lista.Remover(palavra[1]).getName());
        }
    }

    // ---------- Outros Metodos FINAL -----------//

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        // Personagem personagem[] = new Personagem[30];
        ListaPersonagem lista = new ListaPersonagem();

        String ID = sc.nextLine();

        // while (sc.hasNextLine()) {
        while (ID.equals("FIM") != true) {
            lista.InserirFim(ler(ID));
            // personagem[i].imprimir();
            ID = sc.nextLine();
        }

        int numEntradas = sc.nextInt();

        for (int j = 0; j <= numEntradas; j++) {
            ID = sc.nextLine(); // ID = Linha
            hubFuncao(ID, lista);
        }

        lista.imprimir();
        sc.close();
    }
}
// cls && javac Personagem.java && java Personagem < pub.in > saida.txt